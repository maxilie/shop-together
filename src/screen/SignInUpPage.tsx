import { UserData } from '../type/UserData';
import { Keyboard, TouchableWithoutFeedback, View } from 'react-native';
import { ScreenSizeProp, withScreenSize } from '../context/ScreenSizeContext';
import React, { useCallback, useState } from 'react';

// @ts-ignore
import { logo } from '../../assets/logo.png';
import { getDummyPost } from '../constant/DummyData';
import { useSafeAreaInsets } from 'react-native-safe-area-context';
import { createMaterialTopTabNavigator } from '@react-navigation/material-top-tabs';
import MaterialTopTabBarWrapper from '../component/tab/MaterialTopTabBarWrapper';
import SignIn from '../component/login/SignIn';
import SignUp from '../component/login/SignUp';
import { LOGIN_STYLES } from '../component/login/LoginStyles';
import { useFelaNative } from '../hook/UseFelaNative';
import useIsReady from '../hook/ReadyDelay';
import BusyIndicator from '../component/BusyIndicator';

const SignInUpNavigator = createMaterialTopTabNavigator();

type Props = {
  setUserData: (userData: UserData) => void;
}

const SignInUpPage = (props: Props & ScreenSizeProp) => {

  // Initialize state
  const [username, setUsername,] = useState('');
  const [password, setPassword,] = useState('');
  const [passwordConfirm, setPasswordConfirm,] = useState('');
  const [selectedImage, setSelectedImage,] = useState(null);

  // Create styles
  const styleProps = {
    ...LOGIN_STYLES.getStyleProps(props.screenSize),
    insetTop: useSafeAreaInsets().top,
  };
  const style = useFelaNative(styleProps);
  const styles = LOGIN_STYLES.styles;

  // Define function to create user data
  const createUserData = useCallback((): UserData => {
    return {
      name: username || 'Test User',
      posts: [
        getDummyPost(),
        getDummyPost(),
        getDummyPost(),
        getDummyPost(),
        getDummyPost(),
        getDummyPost(),
      ],
      isFollowed: false,
      image: selectedImage !== null
        ? { uri: selectedImage, }
        : require('../../assets/default_profile.png'),
    };
  }, [username, selectedImage,]);

  // Show loading wheel for at least half a second
  // If this is not used, then there could be a delay after user logs out on a slow device
  const isReady = useIsReady();
  if (!isReady) return <BusyIndicator />;

  // Render component
  return (
    <>
      {/**
       * Tap-to-dismiss-keyboard area
       */}
      <TouchableWithoutFeedback
        onPress={Keyboard.dismiss}
        accessible={false}>
        <View
          style={style(styles.tapToDismissKeyboard)} />
      </TouchableWithoutFeedback>

      {/**
       * Tab buttons area
       */}
      <SignInUpNavigator.Navigator
        style={style(styles.navigatorContainer)}
        sceneContainerStyle={style(styles.screenContainer)}
        screenOptions={{
          tabBarPressOpacity: 0.8,
          tabBarContentContainerStyle: style(styles.tabBarContainer),
          tabBarLabelStyle: style(styles.tabBarText),
          tabBarStyle: style(styles.tabBarTextStyle),
        }}
        initialRouteName='SignIn'

        // Setting initial width supposedly aids performance
        initialLayout={{ width: styleProps.screenWidth, }}
        tabBar={MaterialTopTabBarWrapper}
      >

        {/**
         * 'Sign In' tab
         */}
        <SignInUpNavigator.Screen
          name='SignIn'
          options={{
            title: 'Sign In',
          }}
        >
          {(navProps) =>
            <SignIn
              setUserData={props.setUserData}
              username={username}
              setUsername={setUsername}
              password={password}
              setPassword={setPassword}
              createUserData={createUserData}
              {...navProps}
            />}
        </SignInUpNavigator.Screen>

        {/**
         * 'Sign Up' tab
         */}
        <SignInUpNavigator.Screen
          name='SignUp'
          options={{
            title: 'Sign Up',
          }}
        >
          {(navProps) =>
            <SignUp
              setUserData={props.setUserData}
              username={username}
              setUsername={setUsername}
              password={password}
              setPassword={setPassword}
              passwordConfirm={passwordConfirm}
              setPasswordConfirm={setPasswordConfirm}
              selectedImage={selectedImage}
              setSelectedImage={setSelectedImage}
              createUserData={createUserData}
              {...navProps}
            />}
        </SignInUpNavigator.Screen>
      </SignInUpNavigator.Navigator>
    </>
  );
};

export default withScreenSize(SignInUpPage);
