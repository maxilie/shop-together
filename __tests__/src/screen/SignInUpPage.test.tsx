import renderer from 'react-test-renderer';
import { NavigationContainer } from '@react-navigation/native';
import { RendererProvider } from 'react-fela';
import { createRenderer } from 'fela-native';
import SignInUpPage from '../../../src/screen/SignInUpPage';

// Mock nanoid func used for key generation by react-navigation
jest.mock('nanoid/non-secure', () => ({
  nanoid: jest.fn(() => 'mock-nanoid-id'),
}));

describe('SignInUpPage component', () => {
  it('renders correctly', () => {
    const felaRenderer = createRenderer();
    const tree = renderer.create(
      <RendererProvider renderer={felaRenderer}>
        <NavigationContainer>
          <SignInUpPage />
        </NavigationContainer>
      </RendererProvider>).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
