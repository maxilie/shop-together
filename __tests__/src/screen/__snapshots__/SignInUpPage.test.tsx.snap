// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SignInUpPage component renders correctly 1`] = `
Array [
  <View
    accessible={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "height": 828,
        "position": "absolute",
        "width": 1792,
      }
    }
  />,
  <View
    onLayout={[Function]}
    style={
      Array [
        Object {
          "flex": 1,
          "overflow": "hidden",
        },
        Object {
          "backgroundColor": "white",
          "marginTop": 49.68,
          "overflow": "visible",
        },
      ]
    }
  >
    <View
      entering={BaseAnimationMock {}}
      exiting={BaseAnimationMock {}}
      style={
        Object {
          "zIndex": 2,
        }
      }
    >
      <View
        collapsable={false}
        onLayout={[Function]}
        style={
          Object {
            "backgroundColor": "#fafbfb",
            "elevation": 4,
            "shadowColor": "#e1e9ed",
            "shadowOffset": Object {
              "height": 0.9108,
              "width": 0,
            },
            "shadowOpacity": 0.6,
            "shadowRadius": 5.464799999999999,
            "zIndex": 1,
          }
        }
      >
        <View
          collapsable={false}
          pointerEvents="none"
          style={
            Object {
              "bottom": 0,
              "left": 0,
              "position": "absolute",
              "right": 0,
              "top": 0,
            }
          }
        >
          <View
            collapsable={false}
            style={
              Object {
                "backgroundColor": "rgb(0, 122, 255)",
                "bottom": 0,
                "height": 2,
                "left": "0%",
                "position": "absolute",
                "right": 0,
                "transform": Array [],
                "width": "NaN%",
              }
            }
          />
        </View>
        <View
          style={
            Object {
              "overflow": "scroll",
            }
          }
        >
          <RCTScrollView
            accessibilityRole="tablist"
            alwaysBounceHorizontal={false}
            automaticallyAdjustContentInsets={false}
            contentContainerStyle={
              Array [
                Object {
                  "flexDirection": "row",
                  "flexWrap": "nowrap",
                },
                Object {
                  "flex": 1,
                },
                Object {
                  "borderColor": "#f0f4f7",
                  "borderTopWidth": 0.9108,
                  "height": 45.54,
                },
              ]
            }
            data={
              Array [
                Object {
                  "key": "SignIn-mock-nanoid-id",
                  "name": "SignIn",
                  "params": undefined,
                },
                Object {
                  "key": "SignUp-mock-nanoid-id",
                  "name": "SignUp",
                  "params": undefined,
                },
              ]
            }
            getItem={[Function]}
            getItemCount={[Function]}
            horizontal={true}
            initialNumToRender={10}
            keyExtractor={[Function]}
            keyboardShouldPersistTaps="handled"
            onContentSizeChange={[Function]}
            onLayout={[Function]}
            onMomentumScrollBegin={[Function]}
            onMomentumScrollEnd={[Function]}
            onScroll={[Function]}
            onScrollBeginDrag={[Function]}
            onScrollEndDrag={[Function]}
            onViewableItemsChanged={[Function]}
            overScrollMode="never"
            removeClippedSubviews={false}
            renderItem={[Function]}
            scrollEventThrottle={16}
            scrollsToTop={false}
            showsHorizontalScrollIndicator={false}
            showsVerticalScrollIndicator={false}
            stickyHeaderIndices={Array []}
            viewabilityConfigCallbackPairs={
              Array [
                Object {
                  "onViewableItemsChanged": [Function],
                  "viewabilityConfig": undefined,
                },
              ]
            }
          >
            <View>
              <View
                onLayout={[Function]}
                style={
                  Array [
                    Object {
                      "flexDirection": "row",
                    },
                    null,
                  ]
                }
              >
                <View
                  accessibilityRole="tab"
                  accessibilityState={
                    Object {
                      "selected": true,
                    }
                  }
                  accessibilityStates={
                    Array [
                      "selected",
                    ]
                  }
                  accessible={true}
                  collapsable={false}
                  delayPressIn={0}
                  focusable={true}
                  onBlur={[Function]}
                  onClick={[Function]}
                  onFocus={[Function]}
                  onLayout={[Function]}
                  onResponderGrant={[Function]}
                  onResponderMove={[Function]}
                  onResponderRelease={[Function]}
                  onResponderTerminate={[Function]}
                  onResponderTerminationRequest={[Function]}
                  onStartShouldSetResponder={[Function]}
                  style={
                    Array [
                      Object {
                        "opacity": 1,
                      },
                      Array [
                        Object {
                          "backgroundColor": "transparent",
                        },
                        Object {
                          "width": 0,
                        },
                      ],
                    ]
                  }
                >
                  <View
                    pointerEvents="none"
                    style={
                      Array [
                        Object {
                          "alignItems": "center",
                          "flex": 1,
                          "justifyContent": "center",
                          "minHeight": 48,
                          "padding": 10,
                        },
                        undefined,
                      ]
                    }
                  >
                    <View>
                      <View
                        collapsable={false}
                        style={
                          Object {
                            "opacity": 0,
                          }
                        }
                      >
                        <Text
                          style={
                            Array [
                              Object {
                                "backgroundColor": "transparent",
                                "fontSize": 13,
                                "margin": 4,
                                "textAlign": "center",
                                "textTransform": "uppercase",
                              },
                              Object {
                                "color": "rgba(28, 28, 30, 0.5)",
                              },
                              Object {
                                "fontSize": 40.39992079200156,
                                "fontWeight": "bold",
                              },
                            ]
                          }
                        >
                          Sign In
                        </Text>
                      </View>
                      <View
                        collapsable={false}
                        style={
                          Object {
                            "bottom": 0,
                            "left": 0,
                            "opacity": 1,
                            "position": "absolute",
                            "right": 0,
                            "top": 0,
                          }
                        }
                      >
                        <Text
                          style={
                            Array [
                              Object {
                                "backgroundColor": "transparent",
                                "fontSize": 13,
                                "margin": 4,
                                "textAlign": "center",
                                "textTransform": "uppercase",
                              },
                              Object {
                                "color": "rgb(28, 28, 30)",
                              },
                              Object {
                                "fontSize": 40.39992079200156,
                                "fontWeight": "bold",
                              },
                            ]
                          }
                        >
                          Sign In
                        </Text>
                      </View>
                    </View>
                  </View>
                </View>
              </View>
              <View
                onLayout={[Function]}
                style={
                  Array [
                    Object {
                      "flexDirection": "row",
                    },
                    null,
                  ]
                }
              >
                <View
                  accessibilityRole="tab"
                  accessibilityState={
                    Object {
                      "selected": false,
                    }
                  }
                  accessibilityStates={Array []}
                  accessible={true}
                  collapsable={false}
                  delayPressIn={0}
                  focusable={true}
                  onBlur={[Function]}
                  onClick={[Function]}
                  onFocus={[Function]}
                  onLayout={[Function]}
                  onResponderGrant={[Function]}
                  onResponderMove={[Function]}
                  onResponderRelease={[Function]}
                  onResponderTerminate={[Function]}
                  onResponderTerminationRequest={[Function]}
                  onStartShouldSetResponder={[Function]}
                  style={
                    Array [
                      Object {
                        "opacity": 1,
                      },
                      Array [
                        Object {
                          "backgroundColor": "transparent",
                        },
                        Object {
                          "width": 0,
                        },
                      ],
                    ]
                  }
                >
                  <View
                    pointerEvents="none"
                    style={
                      Array [
                        Object {
                          "alignItems": "center",
                          "flex": 1,
                          "justifyContent": "center",
                          "minHeight": 48,
                          "padding": 10,
                        },
                        undefined,
                      ]
                    }
                  >
                    <View>
                      <View
                        collapsable={false}
                        style={
                          Object {
                            "opacity": 1,
                          }
                        }
                      >
                        <Text
                          style={
                            Array [
                              Object {
                                "backgroundColor": "transparent",
                                "fontSize": 13,
                                "margin": 4,
                                "textAlign": "center",
                                "textTransform": "uppercase",
                              },
                              Object {
                                "color": "rgba(28, 28, 30, 0.5)",
                              },
                              Object {
                                "fontSize": 40.39992079200156,
                                "fontWeight": "bold",
                              },
                            ]
                          }
                        >
                          Sign Up
                        </Text>
                      </View>
                      <View
                        collapsable={false}
                        style={
                          Object {
                            "bottom": 0,
                            "left": 0,
                            "opacity": 0,
                            "position": "absolute",
                            "right": 0,
                            "top": 0,
                          }
                        }
                      >
                        <Text
                          style={
                            Array [
                              Object {
                                "backgroundColor": "transparent",
                                "fontSize": 13,
                                "margin": 4,
                                "textAlign": "center",
                                "textTransform": "uppercase",
                              },
                              Object {
                                "color": "rgb(28, 28, 30)",
                              },
                              Object {
                                "fontSize": 40.39992079200156,
                                "fontWeight": "bold",
                              },
                            ]
                          }
                        >
                          Sign Up
                        </Text>
                      </View>
                    </View>
                  </View>
                </View>
              </View>
            </View>
          </RCTScrollView>
        </View>
      </View>
    </View>
    <RNCViewPager
      collapsable={false}
      initialPage={0}
      keyboardDismissMode="on-drag"
      layout={
        Object {
          "height": 0,
          "width": 1792,
        }
      }
      layoutDirection="ltr"
      onMoveShouldSetResponderCapture={[Function]}
      onPageScroll={[Function]}
      onPageScrollStateChanged={[Function]}
      onPageSelected={[Function]}
      scrollEnabled={true}
      style={
        Object {
          "flex": 1,
        }
      }
    >
      <View
        collapsable={false}
        style={
          Object {
            "bottom": 0,
            "left": 0,
            "position": "absolute",
            "right": 0,
            "top": 0,
          }
        }
      >
        <View
          accessibilityElementsHidden={false}
          importantForAccessibility="auto"
          style={
            Array [
              Object {
                "flex": 1,
                "overflow": "hidden",
              },
              Object {
                "width": 1792,
              },
              Array [
                Array [
                  Object {
                    "backgroundColor": "rgb(242, 242, 242)",
                  },
                  Object {
                    "backgroundColor": "white",
                  },
                ],
                Object {
                  "bottom": 0,
                  "left": 0,
                  "position": "absolute",
                  "right": 0,
                  "top": 0,
                },
              ],
            ]
          }
        >
          <View
            accessible={false}
            focusable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Object {
                "alignItems": "center",
                "backgroundColor": "#fafbfb",
                "flexDirection": "column",
                "height": 710.0100000000001,
                "justifyContent": "flex-start",
                "width": 1792,
              }
            }
          >
            <Image
              defaultSource={1}
              style={
                Object {
                  "height": 523.7810945273633,
                  "marginBottom": 41.8968,
                  "marginTop": 23.04324,
                  "resizeMode": "contain",
                  "shadowColor": "#5ee352",
                  "shadowOpacity": 0.2,
                  "shadowRadius": 78.5671641791045,
                  "width": 1433.6000000000001,
                }
              }
            />
            <TextInput
              allowFontScaling={false}
              autoCapitalize="none"
              autoComplete="off"
              clearButtonMode="while-editing"
              maxLength={20}
              onChange={[Function]}
              placeholder="Name"
              placeholderTextColor="#464747"
              secureTextEntry={false}
              selectionColor="#404040"
              style={
                Object {
                  "backgroundColor": "#fafbfb",
                  "borderColor": "#5797bd",
                  "borderRadius": 10.4742,
                  "borderWidth": 1.04742,
                  "fontSize": 18.329849999999997,
                  "height": 52.370999999999995,
                  "marginBottom": 0,
                  "marginTop": 0,
                  "padding": 2.61855,
                  "paddingLeft": 10.4742,
                  "shadowColor": "#e3f4fc",
                  "shadowOpacity": 0.9,
                  "shadowRadius": 4.7133899999999995,
                  "width": 314.226,
                }
              }
              value=""
            />
            <TextInput
              allowFontScaling={false}
              autoCapitalize="none"
              autoComplete="off"
              clearButtonMode="while-editing"
              maxLength={20}
              onChange={[Function]}
              placeholder="Password"
              placeholderTextColor="#464747"
              secureTextEntry={true}
              selectionColor="#404040"
              style={
                Object {
                  "backgroundColor": "#fafbfb",
                  "borderColor": "#5797bd",
                  "borderRadius": 10.4742,
                  "borderWidth": 1.04742,
                  "fontSize": 18.329849999999997,
                  "height": 52.370999999999995,
                  "marginBottom": 0,
                  "marginTop": 20.9484,
                  "padding": 2.61855,
                  "paddingLeft": 10.4742,
                  "shadowColor": "#e3f4fc",
                  "shadowOpacity": 0.9,
                  "shadowRadius": 4.7133899999999995,
                  "width": 314.226,
                }
              }
              value=""
            />
            <View
              accessible={true}
              collapsable={false}
              focusable={true}
              onBlur={[Function]}
              onClick={[Function]}
              onFocus={[Function]}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
            >
              <Text
                style={
                  Object {
                    "color": "#404040",
                    "fontSize": 28.27994455440109,
                    "marginTop": 20.9484,
                  }
                }
              >
                Forgot Password?
              </Text>
            </View>
            <View
              style={
                Object {
                  "transform": Array [
                    Object {
                      "scale": AnimatedValue {
                        " __value": 1,
                      },
                    },
                  ],
                }
              }
            >
              <View
                accessible={true}
                collapsable={false}
                focusable={true}
                onBlur={[Function]}
                onClick={[Function]}
                onFocus={[Function]}
                onResponderGrant={[Function]}
                onResponderMove={[Function]}
                onResponderRelease={[Function]}
                onResponderTerminate={[Function]}
                onResponderTerminationRequest={[Function]}
                onStartShouldSetResponder={[Function]}
                style={
                  Object {
                    "alignItems": "center",
                    "backgroundColor": "#4585c4",
                    "borderRadius": 10.99791,
                    "height": 54.989549999999994,
                    "justifyContent": "center",
                    "marginLeft": "auto",
                    "marginRight": "auto",
                    "marginTop": 41.8968,
                    "shadowColor": "#4585c4",
                    "shadowOffset": Object {
                      "height": 4.399164,
                      "width": 0,
                    },
                    "shadowOpacity": 0.17,
                    "shadowRadius": 16.496864999999996,
                    "width": 1433.6000000000001,
                  }
                }
              >
                <Text
                  style={
                    Object {
                      "color": "white",
                      "fontSize": 27.494774999999997,
                      "fontWeight": "bold",
                    }
                  }
                >
                  SIGN IN
                </Text>
              </View>
            </View>
            <View
              accessible={true}
              collapsable={false}
              focusable={true}
              onBlur={[Function]}
              onClick={[Function]}
              onFocus={[Function]}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
            >
              <Text
                style={
                  Array [
                    Object {
                      "color": "#1084b5",
                      "fontSize": 34.339932673201325,
                      "fontWeight": "bold",
                      "marginTop": 25.13808,
                    },
                    null,
                  ]
                }
              >
                Login with Face ID
              </Text>
            </View>
          </View>
        </View>
      </View>
      <View
        collapsable={false}
        style={
          Object {
            "bottom": 0,
            "left": 0,
            "position": "absolute",
            "right": 0,
            "top": 0,
          }
        }
      >
        <View
          accessibilityElementsHidden={true}
          importantForAccessibility="no-hide-descendants"
          style={
            Array [
              Object {
                "flex": 1,
                "overflow": "hidden",
              },
              Object {
                "width": 1792,
              },
              Array [
                Array [
                  Object {
                    "backgroundColor": "rgb(242, 242, 242)",
                  },
                  Object {
                    "backgroundColor": "white",
                  },
                ],
                Object {
                  "bottom": 0,
                  "left": 0,
                  "position": "absolute",
                  "right": 0,
                  "top": 0,
                },
              ],
            ]
          }
        />
      </View>
    </RNCViewPager>
  </View>,
]
`;
